runtime/src/mm/cpp/ObjectFactory.hpp

```diff
@@ -0,0 +1,317 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+#ifndef RUNTIME_MM_OBJECT_FACTORY_H
+#define RUNTIME_MM_OBJECT_FACTORY_H
+
+#include <algorithm>
+#include <memory>
+#include <mutex>
+
+#include "Alignment.hpp"
+#include "Alloc.h"
+#include "CppSupport.hpp"
+#include "Memory.h"
+#include "Mutex.hpp"
+#include "Utils.hpp"
+
+namespace kotlin {
+namespace mm {
+
+namespace internal {
+
+// A queue that is constructed by collecting subqueues from several `Producer`s.
+// This is essentially a heterogeneous `MultiSourceQueue` on top of a singly linked list that
+// uses `konanAllocMemory` and `konanFreeMemory`
+// TODO: Consider merging with `MultiSourceQueue` somehow.
+template <size_t DataAlignment>
+class ObjectFactoryStorage : private Pinned {
+    static_assert(IsValidAlignment(DataAlignment), "DataAlignment is not a valid alignment");
+
+public:
+    // This class does not know its size at compile-time.
+    class Node : private Pinned {
+        constexpr static size_t DataOffset() noexcept { return AlignUp(sizeof(Node), DataAlignment); }
+
+    public:
+        ~Node() = default;
+
+        static void operator delete(void* ptr) noexcept { konanFreeMemory(ptr); }
+
+        // Note: This can only be trivially destructible data, as nobody can invoke its destructor.
+        void* Data() noexcept {
+            constexpr size_t kDataOffset = DataOffset();
+            void* ptr = reinterpret_cast<uint8_t*>(this) + kDataOffset;
+            RuntimeAssert(IsAligned(ptr, DataAlignment), "Data=%p is not aligned to %zu", ptr, DataAlignment);
+            return ptr;
+        }
+
+        // It's a caller responsibility to know if the underlying data is `T`.
+        template <typename T>
+        T& Data() noexcept {
+            return *static_cast<T*>(Data());
+        }
+
+    private:
+        friend class ObjectFactoryStorage;
+
+        Node() noexcept = default;
+
+        static void* operator new(size_t size, size_t dataSize) noexcept {
+            size_t dataSizeAligned = AlignUp(dataSize, DataAlignment);
+            size_t totalAlignment = std::max(alignof(Node), DataAlignment);
+            size_t totalSize = AlignUp(sizeof(Node) + dataSizeAligned, totalAlignment);
```

Comment:
Why do we need to `AlignUp` when computing the total size? Isn't it enough to call

```
konanAllocAlignedMemory(sizeof(Node) + dataSizeAligned, totalAlignment);
```
instead?

Reply:
Mimicking c++ invariant that `sizeof(T)` is always aligned to `alignof(T)`.